### ISSUE TITLE
Build Issues on Ubuntu 16.04

### ISSUE THREAD

**Issue by eozturk1** [intel/cryptography-primitives#19]
I have the following setup:

```
gcc-8 (Ubuntu 8.4.0-1ubuntu1~16.04.1) 8.4.0
g++-8 (Ubuntu 8.4.0-1ubuntu1~16.04.1) 8.4.0
NASM version 2.15rc0-20191023 compiled on Oct 23 2019
```

NASM was installed from [this](https://www.nasm.us/pub/nasm/snapshots/20191023/) 2.15 snapshot as suggested -- although I had to convert the `rpm` file to `deb` with [alien](https://wiki.debian.org/Alien) tool.

Error messages I receive are:
```
[  4%] Built target ippcp_dyn_mx-asmobjs
[  4%] Building C object sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/cpinit.c.o
/tmp/cc5yldCq.s: Assembler messages:
/tmp/cc5yldCq.s:289: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:301: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:334: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:350: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:362: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:537: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:549: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:567: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:604: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:632: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:651: Error: no such instruction: `endbr64'
/tmp/cc5yldCq.s:685: Error: no such instruction: `endbr64'
sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/build.make:62: recipe for target 'sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/cpinit.c.o' failed
make[2]: *** [sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/cpinit.c.o] Error 1
CMakeFiles/Makefile2:101: recipe for target 'sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/all' failed
make[1]: *** [sources/ippcp/CMakeFiles/ippcp_dyn_mx.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```

It seems the problem is with the `endbr64` instruction. I tried changing the platform list to `mx` but that gives the same error.

Is there any way to fix this? Is it not possible to compile the library on Ubuntu 16.04?

I also searched for static library files but could not find them. Are they available anywhere?

**Comment by amatyuko-intc** [comment]
Please make sure you have up-to-date binutils package. I suggest to install binutils 2.32 or newer.

Regards,
Andrey

**Comment by eozturk1** [comment]
It seems binutils 2.32 is not available for Ubuntu 16.04 -- current version they have available is 2.26.1-1ubuntu1~16.04.8. Installing it manually will take me down the following dependency rabbit hole:

```
binutils-common (= 2.32-7ubuntu1)
binutils-x86-64-linux-gnu (= 2.32-7ubuntu1)
libbinutils (= 2.32-7ubuntu1)
libc6 (>= 2.27)
zlib1g (>= 1:1.1.4)
```

Is there an automatic way to install all of them? 

Are there static libraries available for IPP-Crypto?

**Comment by amatyuko-intc** [comment]
You could try to build them (latest binutils) from sources to use on Ubuntu 16.04.

Alternatively, doing it on your own risk, you can also disable CET-protection (essentially usage of endbr instructions) by the [patch](https://github.com/intel/ipp-crypto/files/4696217/disable_cet.patch.txt) attached.

The static pre-built libraries are now available upon request only, please see this link for details: https://software.intel.com/content/www/us/en/develop/articles/get-ipp-cryptography-libraries.html

Regards,
Andrey

**Comment by eozturk1** [comment]
I didn't know disabling CET-protection was an option. Thank you for sharing this patch!

Eventually we went with the route of getting a new machine with Ubuntu 20.04 which already had the required libraries in available repositories -- except NASM 2.15 which I had to install from the provided snapshot after converting from `rpm` to `deb` using `alien`.

I tried installing the pre-built libraries but for some reason the installation process claimed that it did not have an active Internet connection even though I tried it multiple times -- maybe a firewall issue blocked access?.
